packages/foam-vscode/src/features/orphans.ts

```diff
@@ -0,0 +1,130 @@
+import * as path from 'path';
+import * as vscode from 'vscode';
+import { Foam, Note } from 'foam-core';
+import { getOrphansConfig, OrphansConfig } from '../settings';
+import { FoamFeature } from '../types';
+
+const feature: FoamFeature = {
+  activate: async (
+    context: vscode.ExtensionContext,
+    foamPromise: Promise<Foam>
+  ) => {
+    const foam = await foamPromise;
+    const config = getOrphansConfig();
+    const provider = new OrphansProvider(foam, config);
+
+    context.subscriptions.push(
+      vscode.window.registerTreeDataProvider('foam-vscode.orphans', provider)
+    );
+
+    foam.notes.onDidUpdateNote(() => provider.refresh());
+  },
+};
+
+export default feature;
+
+class OrphansProvider implements vscode.TreeDataProvider<OrphanTreeItem> {
+  // prettier-ignore
+  private _onDidChangeTreeData: vscode.EventEmitter<OrphanTreeItem | undefined | void> = new vscode.EventEmitter<OrphanTreeItem | undefined | void>();
+  // prettier-ignore
+  readonly onDidChangeTreeData: vscode.Event<OrphanTreeItem | undefined | void> = this._onDidChangeTreeData.event;
+
+  private orphans: Note[] = [];
+  private exclude: string[] = [];
+  private root = vscode.workspace.workspaceFolders[0].uri.fsPath;
+
+  constructor(private foam: Foam, private config: OrphansConfig) {
+    this.exclude = config.exclude.map(d => path.normalize(`/${d}`));
+    this.computeOrphans();
+  }
+
+  refresh(): void {
+    this.computeOrphans();
+    this._onDidChangeTreeData.fire();
+  }
+
+  private computeOrphans() {
+    this.orphans = this.foam.notes
+      .getNotes()
+      .filter(note => !this.foam.notes.getAllLinks(note.uri).length)
+      .filter(note => {
+        const p = note.uri.path.replace(this.root, '');
+        const { dir } = path.parse(p);
+        return !this.exclude.includes(dir);
+      })
+      .sort((a, b) => a.title.localeCompare(b.title));
+  }
+
+  private getOrphansByDirectory(): OrphansByDirectory {
+    const orphans: OrphansByDirectory = {};
+    for (const orphan of this.orphans) {
+      const p = orphan.uri.path.replace(this.root, '');
+      const { dir } = path.parse(p);
+
+      if (orphans[dir]) {
+        orphans[dir].push(orphan);
+      } else {
+        orphans[dir] = [orphan];
+      }
+    }
+
+    for (const k in orphans) {
+      orphans[k].sort((a, b) => a.title.localeCompare(b.title));
+    }
+
+    return orphans;
+  }
+
+  getTreeItem(item: OrphanTreeItem): vscode.TreeItem {
+    return item;
+  }
+
+  getChildren(directory?: Directory): Thenable<OrphanTreeItem[]> {
+    if (!directory && this.config.groupBy === 'folder') {
```

Comment:
IIRC unfortunately in VsCode it's not possible for extensions to have buttons in their treeview.
It might be worth double checking because it would be amazing to be able to flip the setting here

![image](https://user-images.githubusercontent.com/457005/105583923-9320b480-5d90-11eb-92f3-13a5f908eb81.png)

Reply:
@riccardoferretti actually I managed to add see, see the incoming coming and the gif in the PR description

Reply:
That's great!

A few notes:
- what's your take on the best way to manage the setting vs the switch?
- can we use the same icons that VsCode already uses in the Source Control panel? https://github.com/microsoft/vscode-extension-samples/tree/master/webview-codicons-sample (this example is for webview, but might be a good starting point to find the needed info)
- is there a way to not show the switch command in the command palette? I can't imagine someone needing it there (but not sure whether VsCode allows for that)

![image](https://user-images.githubusercontent.com/457005/105626827-d29fdd00-5e32-11eb-87f8-1e5f7e01f610.png)
![image](https://user-images.githubusercontent.com/457005/105626830-da5f8180-5e32-11eb-8521-4d0932b1fd20.png)

Reply:
@riccardoferretti 

> what's your take on the best way to manage the setting vs the switch?

I meant to ask your take on that too. I see two solutions:
- either keep both switch and setting; the setting is the default for the next time the repo is opened
- discard the setting completely and rely uniquely on the switch

Will look at the other points later, but I guess it's possible maybe

Reply:
I agree, the one thing I want to avoid is the switch affecting the settings, which would be too noisy. 

I am AFK now, but I am curious to see if the source control panel remembers the last state, and if so, how.

I am comfortable with either approach, as we already support the setting let's keep it, and make it the default state for the panel.
For people who don't care, they won't have the setting defined and this would just work. For people who care, they will be able to adjust it.

Reply:
@riccardoferretti coming back on the two other points:

- it is possible to hide the groupby toggle command from the palette, see latest commit
- I went through quite a few docs but I don't think it's possible to have a toggle between two icons in an extension-contributed menu, if you know of any extension that has this let me know and I'll check how they did it. What we can do now is use either the flat list or tree list icon, but no toggle. Below is a more detailed explanation of my findings

---

First, it seems the toggle from the SCM view is baked into vscode, so extensions cannot reuse the same mechanism:

https://github.com/microsoft/vscode/blob/d4be66da200acc87e01d3517bfddf799fd8d110b/src/vs/workbench/contrib/scm/browser/scmViewPane.ts#L1208-L1217

Then, what I though would do the trick would be to register the button for the toggle command two times, with a different set of icons and a different command, and hide one of them using the `when` clause depending on state. Something like this:

```json
    "menus": {
      "view/title": [
        {
          "command": "foam-vscode.toggle-orphans-list-tree",
          "when": "view == foam-vscode.orphans && someState == flat", // here manage the state
          "group": "navigation"
        },
        {
          "command": "foam-vscode.toggle-orphans-list-flat",
          "when": "view == foam-vscode.orphans && someState == tree", // here manage the state
          "group": "navigation"
        }
      ],
      "commandPalette": [
        {
          "command": "foam-vscode.toggle-orphans-list-tree",
          "when": "false"
        },
        {
          "command": "foam-vscode.toggle-orphans-list-flat",
          "when": "false"
        }
      ]
    },
    "commands": [
      {
        "command": "foam-vscode.toggle-orphans-list-tree",
        "title": "Foam: Display orphans grouped by folder",
        "icon": {
          "light": "static/icons/light/list-tree.svg",
          "dark": "static/icons/dark/list-tree.svg"
        }
      },
      {
        "command": "foam-vscode.toggle-orphans-list-flat",
        "title": "Foam: Display all orphans",
        "icon": {
          "light": "static/icons/light/list-flat.svg",
          "dark": "static/icons/dark/list-flat.svg"
        }
      }
    ],
```

Thing is, I don't think we can manage the state with the `when` clause, as we can't access any state managed by the extension from there: https://code.visualstudio.com/docs/getstarted/keybindings#_when-clause-contexts Or at least I don't see any way yet

What it can do is, it can read from the whole `config` object, but I think that'd mean we'd overwrite the user defined setting for `groupBy` :thinking: and that's not what we want.

If we could access `globalState` or `workspaceState` from the `when` clause though we'd be golden and could have the toggle: https://code.visualstudio.com/api/references/vscode-api#ExtensionContext

Reply:
I have seen that this extension (https://github.com/Gruntfuggly/todo-tree/) is achieving the button switch, and here is the approach they seem to be taking: https://github.com/Gruntfuggly/todo-tree/search?q=todo-tree-flat - what do you think?

Reply:
That looks like exactly what we needed, seems like the docs for this were well hidden: https://code.visualstudio.com/api/extension-guides/command#using-a-custom-when-clause-context

Reply:
@riccardoferretti Should be all good now! Just pushed the toggle and added a GIF